src/ngraph/node.cpp

```diff
@@ -46,32 +46,47 @@ Node::Node(const std::string& node_type, const NodeVector& arguments)
             m_inputs.emplace_back(this, i++, output);
         }
     }
+    set_output_size(output_size);
 }
 
-void Node::set_value_type_checked(const element::Type& element_type, const Shape& shape)
+static bool in_transition = true;
+
+void Node::constructor_validate_and_infer_types()
 {
-    if (m_outputs.size() == 0)
+    if (in_transition)
     {
-        add_output(element_type, shape);
+        validate_and_infer_types();
     }
-    if (element_type != get_element_type() || shape != get_shape())
+}
+
+void Node::delayed_validate_and_infer_types()
+{
+    if (!in_transition)
+    {
+        validate_and_infer_types();
+    }
+}
+
+void Node::set_output_size(size_t n)
+{
+    m_outputs.clear();
```

Comment:
wonder if it makes to first check if for all outputs the element type is `unspecified` meaning this is the first time running this function. This way we could try catching any situations resulting in overriding already set types?

Reply:
If validate... gets called a second time (like after a pass), it wants to reset things.
It would be nice to have a way to just do the new nodes spliced in, but that seems hard to do without a lot of error-prone checks in every op.